src/components/autocomplete/js/autocompleteController.js

```diff
@@ -0,0 +1,145 @@
+(function () {
+  'use strict';
+  angular
+      .module('material.components.autocomplete')
+      .controller('MdAutocompleteCtrl', MdAutocompleteCtrl);
+
+  function MdAutocompleteCtrl ($scope, $element, $timeout, $q, $mdUtil) {
+
+    //-- private variables
+    var self = this,
+        itemParts = $scope.itemsExpr.split(/\ in\ /i),
+        itemExpr = itemParts[1],
+        elements = {
+          main:  $element[0],
+          ul:    $element[0].getElementsByTagName('ul')[0],
+          input: $element[0].getElementsByTagName('input')[0]
+        },
+        promise  = null,
+        cache    = {};
+
+    //-- public variables
+    self.itemName = itemParts[0];
+    self.matches  = [];
+    self.loading  = false;
+    self.hidden   = true;
+    self.index    = 0;
+    self.keydown  = keydown;
+    self.clear    = clearValue;
+    self.select   = select;
+    self.fetch    = $mdUtil.debounce(fetchResults);
+
+    //-- return init
+    return init();
+
+    //-- start method definitions
+    function init () {
+      configureWatchers();
+      attachAriaAttributes();
+    }
+
+    function attachAriaAttributes () {
+      var ul = angular.element(elements.ul),
+          input = angular.element(elements.input),
+          id = ul.attr('id') || 'ul_' + $mdUtil.nextUid();
+      ul.attr('id', id);
+      input.attr('aria-owns', id);
+    }
+
+    function configureWatchers () {
+      $scope.$watch('searchText', function (searchText) {
+        if (!searchText) {
+          self.loading = false;
+          return self.matches = [];
+        }
+        var term = searchText.toLowerCase();
+        if (promise && promise.cancel) {
+          promise.cancel();
+          promise = null;
+        }
+        if (cache[term]) {
+          self.matches = cache[term];
+        } else if (!self.hidden) {
+          self.loading = true;
+          self.fetch(searchText);
+        }
+      });
+    }
+
+    function fetchResults (searchText) {
+      var items = $scope.$parent.$eval(itemExpr),
+          term = searchText.toLowerCase();
+      promise = $q.when(items).then(function (matches) {
+        cache[term] = matches;
+        if (searchText !== $scope.searchText) return; //-- just cache the results if old request
+        promise = null;
+        self.loading = false;
+        self.matches = matches;
+      });
+    }
+
+    function keydown (event) {
+      switch (event.keyCode) {
+        case 40: //-- DOWN ARROW
```

Comment:
These keyCodes are all stored in `$mdConstant`. You could use `$mdConstant.KEYCODE.DOWN_ARROW`.